src/main/steps/applicant/statement-of-service/choose-parties/content.ts

```diff
@@ -0,0 +1,247 @@
+          classes: 'govuk-input--width-4',
+          attributes: { maxLength: 4, pattern: '[0-9]*', inputMode: 'numeric' },
+        },
+      ],
+      parser: body => covertToDateObject('partiesServedDate', body as Record<string, unknown>),
+      validator: value =>
+        areDateFieldsFilledIn(value as CaseDate) ||
+        isDateInputInvalid(value as CaseDate) ||
+        isFutureDate(value as CaseDate),
+    },
+    partiesServed: {
+      type: 'checkboxes',
+      label: l => l.whowasserved,
+      values: [...checkboxes],
+      validator: atLeastOneFieldIsChecked,
+    },
+  };
+
+  return { fields, errors };
+};
+
+export const form: FormContent = {
+  fields: {},
+  onlyContinue: {
+    text: l => l.onlycontinue,
+  },
+};
+
+const getParties = (userCase: Partial<CaseWithId>) => {
+  const parties: { id: string; value: string }[] = [];
+  userCase?.respondents?.forEach(respondent => {
+    if (respondent.value.response.citizenFlags?.isApplicationServed !== 'Yes') {
+      parties.push({
+        id: respondent.id,
+        value: respondent.value.firstName + ' ' + respondent.value.lastName,
+      });
+    }
+  });
+  userCase?.applicants?.forEach(applicant => {
+    if (applicant.value.response.citizenFlags?.isApplicationServed !== 'Yes') {
+      parties.push({
+        id: applicant.id,
+        value: applicant.value.firstName + ' ' + applicant.value.lastName,
+      });
+    }
+  });
+  return parties;
+};
+
+export const prepateStatementOfServiceRequest = (
+  req: AppRequest<AnyObject>,
+  formData: Partial<Case>
+): Partial<CaseWithId> => {
+  const userCase = req.session.userCase;
+  userCase.partiesServed = formData.partiesServed as string[];
+  if (userCase.partiesServed && formData.partiesServedDate) {
+    const date = formData.partiesServedDate as unknown;
+    const date2 = date as { day: string; month: string; year: string };
+    userCase.partiesServed = userCase.partiesServed.filter(party => party !== '');
+
+    userCase.partiesServed.forEach(partyId => markSosFlag(userCase, partyId));
+    if (userCase.applicants && userCase.applicants[0]) {
+      const partyNames = getPartiesSelected(userCase.partiesServed, userCase);
+      if (getParties(userCase).length === 0) {
+        if (userCase.applicants[0].value.response.citizenFlags) {
+          userCase.applicants[0].value.response.citizenFlags.isStatementOfServiceProvided = 'Yes';
+        } else {
+          userCase.applicants[0].value.response.citizenFlags = {
+            isStatementOfServiceProvided: 'Yes',
+          };
+        }
+      }
+      userCase.applicants[0].value.citizenSosObject = {
+        partiesServed: partyNames,
+        partiesServedDate: date2.year + '-' + date2.month + '-' + date2.day,
+        citizenSosDocs: userCase.docIdList,
+      };
+      userCase.docIdList = [];
+    }
+  }
+  return userCase;
+};
+
+const markSosFlag = (userCase: Partial<CaseWithId>, partyId: string) => {
+  userCase?.applicants?.forEach(party => {
```

Comment:
This constructor is costly. Please use a constructor that takes care of performance.